spatialist/raster.py

```diff
@@ -1030,63 +1034,62 @@ def stack(srcfiles, dstfile, resampling, targetres, srcnodata, dstnodata, shapef
     
     if layernames is not None:
         if len(layernames) != len(srcfiles):
-            raise IOError('mismatch between number of source file groups and layernames')
+            raise RuntimeError('mismatch between number of source file groups and layernames')
     
     if not isinstance(targetres, (list, tuple)) or len(targetres) != 2:
         raise RuntimeError('targetres must be a list or tuple with two entries for x and y resolution')
     
     if len(srcfiles) == 1 and not isinstance(srcfiles[0], list):
-        raise IOError('only one file specified; nothing to be done')
+        raise RuntimeError('only one file specified; nothing to be done')
     
-    if resampling not in ['near', 'bilinear', 'cubic', 'cubicspline', 'lanczos', 'average', 'mode', 'max', 'min', 'med',
-                          'Q1', 'Q3']:
-        raise IOError('resampling method not supported')
+    if resampling not in ['near', 'bilinear', 'cubic', 'cubicspline', 'lanczos',
+                          'average', 'mode', 'max', 'min', 'med', 'Q1', 'Q3']:
+        raise RuntimeError('resampling method not supported')
     
     projections = list()
     for x in dissolve(srcfiles):
         try:
             projection = Raster(x).projection
-        except OSError as e:
+        except RuntimeError as e:
             print('cannot read file: {}'.format(x))
             raise e
         projections.append(projection)
     
     projections = list(set(projections))
     if len(projections) > 1:
-        raise IOError('raster projection mismatch')
-    elif len(projections) == 0:
+        raise RuntimeError('raster projection mismatch')
+    elif projections[0] == '':
         raise RuntimeError('could not retrieve the projection from any of the {} input images'.format(len(srcfiles)))
     else:
         srs = projections[0]
     
     # read shapefile bounding coordinates and reduce list of rasters to those overlapping with the shapefile
     if shapefile is not None:
-        shp = shapefile if isinstance(shapefile, Vector) else Vector(shapefile)
+        shp = shapefile.clone() if isinstance(shapefile, Vector) else Vector(shapefile)
         shp.reproject(srs)
         ext = shp.extent
         arg_ext = (ext['xmin'], ext['ymin'], ext['xmax'], ext['ymax'])
-        for i in range(len(srcfiles)):
-            group = sorted(srcfiles[i], key=sortfun) if isinstance(srcfiles[i], list) else [srcfiles[i]]
+        for i, item in enumerate(srcfiles):
+            group = item if isinstance(item, list) else [item]
+            if layernames is None and sortfun is not None:
+                group = sorted(group, key=sortfun)
             group = [x for x in group if intersect(shp, Raster(x).bbox())]
             if len(group) > 1:
                 srcfiles[i] = group
             elif len(group) == 1:
                 srcfiles[i] = group[0]
             else:
                 srcfiles[i] = None
+        shp.close()
```

Comment:
Do I have to close a vector object?

Reply:
Yes. At least it is safer. The Vector object contains a file connection and you want to make sure to delete that connection once done. This is what the with-syntax is for. Any connection opened py e.g. ogr.Open needs to be set to None to close the connection and delete the underlying C++ object. See [here](https://trac.osgeo.org/gdal/wiki/PythonGotchas#CertainobjectscontainaDestroymethodbutyoushouldneveruseit) and [here](https://github.com/johntruckenbrodt/spatialist/blob/96fd2e964a26f806d88d89bce98874d538ec6b97/spatialist/vector.py#L207).